src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs

```diff
@@ -602,8 +602,8 @@ public virtual byte[] ToArray()
             int count = _length - _origin;
             if (count == 0)
                 return Array.Empty<byte>();
-            byte[] copy = new byte[count];
-            Buffer.BlockCopy(_buffer, _origin, copy, 0, count);
+            byte[] copy = GC.AllocateUninitializedArray<byte>(count);
+            _buffer.AsSpan(_origin, count).CopyTo(copy);
             return copy;
```

Comment:
Makes me wonder if `{ReadOnly}Span<T>`'s ToArray should be changed to use `AllocateUninitializedArray`, at which point this whole method becomes:
```C#
public virtual byte[] ToArray() =>
    new ReadOnlySpan<byte>(_buffer, _origin, _length - _origin).ToArray();
```

Reply:
On the whole, I suspect `MemoryStream`-backed buffers tend to be larger than the average span. But that makes me wonder if we could make an optimization in `AllocateUninitializedArray` that does something like:

1. Don't bother checking the array size, just always attempt to allocate from the Gen0 heap. (This would need to be a different helper than the current `newarr` helper, which goes down a "collect and try again" fallback path.)
2. If the allocation succeeds, return the newly created array as-is. Sure, it'll be zeroed-out, but whatever. At least it was O(1) to create.
3. If the allocation fails, go down normal `AllocateUninitializedArray` logic.

I don't know enough about this space to know if this is a good idea.

Reply:
Looking at the usages around System and the spans appear to be fairly short and need to be 2048+ for the optimization to kick in.
It's a good idea and I'm only seeing about 1ns cost due to the extra branch but it feels like AllocateUninitializedArray needs to be an improvement for at least 256+ lengths?

Reply:
I experimented with prototyping the "fast zero-init allocator" as mentioned in my previous comment. Code is at https://github.com/GrabYourPitchforks/runtime/commit/2a61989cc863178870c557bce0191ea03d722e06, though a warning that it's just prototype-quality. This makes `GC.AllocateUninitializedArray<T>` competitive with `new T[]` in terms of throughput.